File: sdk/julia/src/insert.jl
const SAVEPOINT = "sienna_griddb_tools_insert"

function with_savepoint(f::Function, db::SQLite.DB)
    DBInterface.execute(db, "SAVEPOINT $SAVEPOINT")
    try
        f()
    catch
        DBInterface.execute(db, "ROLLBACK TO $SAVEPOINT")
        DBInterface.execute(db, "RELEASE $SAVEPOINT")
        rethrow()
    end
    DBInterface.execute(db, "RELEASE $SAVEPOINT")
    return nothing
end

wrap_error(e::SQLite.SQLiteException, what::AbstractString) = InsertError("$what: $(e.msg)")
wrap_error(e::EncodeError, what::AbstractString) = InsertError("$what: $(e.msg)")
wrap_error(e, ::AbstractString) = e

"""
Prepared statements for one insert call, keyed by SQL text, so each is compiled once.
"""
struct StatementCache
    db::SQLite.DB
    stmts::Dict{String, SQLite.Stmt}
end

StatementCache(db::SQLite.DB) = StatementCache(db, Dict{String, SQLite.Stmt}())

function close_statements!(cache::StatementCache)
    foreach(DBInterface.close!, values(cache.stmts))
    empty!(cache.stmts)
    return nothing
end

function with_statements(f::Function, db::SQLite.DB)
    cache = StatementCache(db)
    try
        with_savepoint(() -> f(cache), db)
    finally
        close_statements!(cache)
    end
    return nothing
end

function run_sql(
    cache::StatementCache,
    sql::AbstractString,
    params::Vector,
    what::AbstractString,
)
    try
        stmt = get!(() -> SQLite.Stmt(cache.db, sql), cache.stmts, sql)
        DBInterface.execute(stmt, params)
    catch e
        throw(wrap_error(e, what))
    end
    return nothing
end

function bound_params(bindings::Vector{Binding}, obj::AbstractDict, what::AbstractString)
    try
        return Any[bound_value(b.encoding, obj, b.segments) for b in bindings]
    catch e
        throw(wrap_error(e, what))
    end
end

function skip_field!(report::InsertReport, strict::Bool, type_name, field, what)
    if strict
        throw(GapValueError("$what: field $(repr(field)) has no column in GridDB"))
    end
    add_skipped!(report, type_name, field)
    return nothing
end

function mark_unsupported!(report::InsertReport, strict::Bool, key, n::Int, reason)
    if strict
        throw(UnsupportedComponentError("$key ($n rows): $reason"))
    end
    add_unsupported!(report, key, n)
    return nothing
end

function unit_columns(attr::AttributePlan)
    if attr.registered
        return Any[attr.unit, attr.quantity_kind]
    end
    return Any[missing, missing]
end

function attribute_params(attr::AttributePlan, plan::ComponentPlan, obj, value)
    return Any[
        obj["id"],
        plan.type_name,
        attr.field,
        canonical_json(value),
        unit_columns(attr)...,
    ]
end

function write_attribute!(cache, report, strict, plan, attr::AttributePlan, obj, what)
    value = get(obj, attr.field, nothing)
    if isnothing(value)
        return nothing
    end
    if !attr.registered && !unit_free_value(value)
        skip_field!(report, strict, plan.type_name, attr.field, what)
        return nothing
    end
    run_sql(cache, manifest().attribute_sql, attribute_params(attr, plan, obj, value), what)
    return nothing
end

function write_row!(cache, plan::ComponentPlan, obj::AbstractDict, report, strict::Bool)
    what = describe(plan.type_name, obj)
    run_sql(cache, plan.row_sql, bound_params(plan.bindings, obj, what), what)
    for attr in plan.attributes
        write_attribute!(cache, report, strict, plan, attr, obj, what)
    end
    for gap in plan.gaps
        if !isnothing(get(obj, gap, nothing))
            skip_field!(report, strict, plan.type_name, gap, what)
        end
    end
    unknown = String[k for k in keys(obj) if !(k in plan.known_fields || isnothing(obj[k]))]
    for key in sort!(unknown)
        skip_field!(report, strict, plan.type_name, key, what)
    end
    add_inserted!(report, plan.type_name)
    return nothing
end

function write_entity!(cache, plan::ComponentPlan, obj::AbstractDict)
    what = describe(plan.type_name, obj)
    run_sql(cache, plan.entity_sql, bound_params(plan.entity_bindings, obj, what), what)
    return nothing
end

function note_unsupported!(report, strict, type_name::AbstractString, n::Int)
    reason =
        get(manifest().unsupported_components, type_name, "not a GridDB component type")
    mark_unsupported!(report, strict, type_name, n, reason)
    return nothing
end

"""
    insert_components!(db, type_name, objs; strict=false) -> InsertReport

Insert JSON-shaped objects of one SDK component type, in one savepoint.
"""
function insert_components!(
    db::SQLite.DB,
    type_name::AbstractString,
    objs::AbstractVector;
    strict::Bool=false,
)
    report = InsertReport()
    components = manifest().components
    if !haskey(components, type_name)
        note_unsupported!(report, strict, type_name, length(objs))
        return report
    end
    plan = components[type_name]
    with_statements(db) do cache
        for obj in objs
            write_entity!(cache, plan, obj)
            write_row!(cache, plan, obj, report, strict)
        end
    end
    return report
end

"""
    insert_component!(db, type_name, obj; strict=false) -> InsertReport
"""
function insert_component!(
    db::SQLite.DB,
    type_name::AbstractString,
    obj::AbstractDict;
    strict::Bool=false,
)
    return insert_components!(db, type_name, [obj]; strict=strict)
end

function attribute_types(doc::AbstractDict)
    types = Dict{Int, String}()
    for assoc in get(doc, "supplemental_attribute_associations", Any[])
        types[Int(assoc["attribute_id"])] = assoc["attribute_type"]
    end
    return types
end

section_size(rows) = length(rows)
section_size(::Nothing) = 0

function supplemental_table(attr_type::AbstractString)
    if attr_type in manifest().plant_types
        return "plants"
    end
    return "supplemental_attributes"
end

function write_supplemental!(cache, table::AbstractString, attr_type, attr, what)
    m = manifest()
    if table == "plants"
        value = Dict(k => v for (k, v) in attr if k != "id" && k != "name")
        params = Any[attr["id"], attr["name"], attr_type, canonical_json(value)]
        run_sql(cache, m.plant_sql, params, what)
        return nothing
    end
    value = Dict(k => v for (k, v) in attr if k != "id")
    params = Any[attr["id"], attr_type, canonical_json(value)]
    run_sql(cache, m.supplemental_attribute_sql, params, what)
    return nothing
end

"""
    insert_document!(db, doc; strict=false) -> InsertReport

Insert a whole `SystemDocument` (as parsed JSON) in one savepoint: vocabulary, every
entity row, typed rows in foreign-key rank order, supplemental attributes, associations.
Any failure rolls back the whole document.
"""
function insert_document!(db::SQLite.DB, doc::AbstractDict; strict::Bool=false)
    m = manifest()
    report = InsertReport()
    components = get(doc, "components", Dict{String, Any}())
    plans = Tuple{ComponentPlan, Vector{Any}}[]
    for type_name in sort!(collect(keys(components)))
        objs = components[type_name]
        if haskey(m.components, type_name)
            push!(plans, (m.components[type_name], collect(Any, objs)))
        else
            note_unsupported!(report, strict, type_name, length(objs))
        end
    end
    sort!(plans; by=p -> (p[1].rank, p[1].type_name))
    for section in sort!(collect(keys(m.unsupported_sections)))
        n = section_size(get(doc, section, nothing))
        if n > 0
            mark_unsupported!(report, strict, section, n, m.unsupported_sections[section])
        end
    end
    attr_types = attribute_types(doc)
    with_statements(db) do cache
        seed_vocabulary!(db)
        for (plan, objs) in plans
            for obj in objs
                write_entity!(cache, plan, obj)
            end
        end
        routed = Tuple{String, String, Any, String}[]
        for attr in get(doc, "supplemental_attributes", Any[])
            id = Int(attr["id"])
            if !haskey(attr_types, id)
                throw(
                    InsertError(
                        "supplemental attribute id=$id: no association names its type",
                    ),
                )
            end
            attr_type = attr_types[id]
            table = supplemental_table(attr_type)
            what = "$attr_type id=$id"
            run_sql(
                cache,
                "INSERT INTO entities (id, entity_table, entity_type) VALUES (?, ?, ?)",
                Any[id, table, attr_type],
                what,
            )
            push!(routed, (table, attr_type, attr, what))
        end
        for (plan, objs) in plans
            for obj in objs
                write_row!(cache, plan, obj, report, strict)
            end
        end
        for (table, attr_type, attr, what) in routed
            write_supplemental!(cache, table, attr_type, attr, what)
            add_inserted!(report, table)
        end
        for section in m.associations
            for row in get(doc, section.section, Any[])
                what = "$(section.section) row $(JSON.json(row))"
                run_sql(
                    cache,
                    section.row_sql,
                    bound_params(section.bindings, row, what),
                    what,
                )
                add_inserted!(report, section.section)
            end
        end
    end
    return report
end

File: sdk/typescript/src/insert.ts
import Database from "better-sqlite3";
import type { Connection } from "./db.js";
import { seedVocabulary } from "./db.js";
import { canonicalJson, encode, isNull, valueAt, type SqlValue } from "./encode.js";
import {
  EncodeError,
  GapValueError,
  InsertError,
  UnsupportedComponentError,
  describe,
  type JsonObject,
} from "./errors.js";
import { loadManifest, type Binding, type ComponentPlan } from "./manifest.js";
import { InsertReport } from "./report.js";

const SAVEPOINT = "sienna_griddb_tools_insert";

export interface InsertOptions { strict?: boolean }

function withSavepoint(db: Connection, f: () => void): void {
  db.exec(`SAVEPOINT ${SAVEPOINT}`);
  try {
    f();
  } catch (err) {
    db.exec(`ROLLBACK TO ${SAVEPOINT}`);
    db.exec(`RELEASE ${SAVEPOINT}`);
    throw err;
  }
  db.exec(`RELEASE ${SAVEPOINT}`);
}

function wrap(err: unknown, what: string): unknown {
  if (err instanceof EncodeError || err instanceof Database.SqliteError) {
    return new InsertError(`${what}: ${err.message}`);
  }
  return err;
}

// better-sqlite3 does not cache prepared statements; compile each SQL text once per connection.
const statements = new WeakMap<Connection, Map<string, Database.Statement>>();

function prepared(db: Connection, sql: string): Database.Statement {
  let cache = statements.get(db);
  if (!cache) {
    cache = new Map();
    statements.set(db, cache);
  }
  let stmt = cache.get(sql);
  if (!stmt) {
    stmt = db.prepare(sql);
    cache.set(sql, stmt);
  }
  return stmt;
}

function run(db: Connection, sql: string, params: SqlValue[], what: string): void {
  try {
    prepared(db, sql).run(...params);
  } catch (err) {
    throw wrap(err, what);
  }
}

function params(bindings: Binding[], obj: JsonObject, what: string): SqlValue[] {
  try {
    return bindings.map((b) => encode(b.encode, valueAt(obj, b.path)));
  } catch (err) {
    throw wrap(err, what);
  }
}

function skip(report: InsertReport, strict: boolean, typeName: string, field: string, what: string) {
  if (strict) {
    throw new GapValueError(`${what}: field ${JSON.stringify(field)} has no column in GridDB`);
  }
  report.addSkipped(typeName, field);
}

function unsupported(report: InsertReport, strict: boolean, key: string, n: number, reason: string) {
  if (strict) throw new UnsupportedComponentError(`${key} (${n} rows): ${reason}`);
  report.addUnsupported(key, n);
}

function writeRow(
  db: Connection,
  plan: ComponentPlan,
  obj: JsonObject,
  report: InsertReport,
  strict: boolean,
) {
  const what = describe(plan.typeName, obj);
  run(db, plan.row_sql, params(plan.bindings, obj, what), what);
  for (const attr of plan.attributes) {
    const value = obj[attr.field];
    if (isNull(value)) continue;
    if (attr.unit === null && typeof value !== "string" && typeof value !== "boolean") {
      skip(report, strict, plan.typeName, attr.field, what);
      continue;
    }
    run(
      db,
      loadManifest().attribute_sql,
      [BigInt(obj.id as number), plan.typeName, attr.field, canonicalJson(value), attr.unit, attr.quantity_kind],
      what,
    );
  }
  for (const gap of plan.gaps) {
    if (!isNull(obj[gap])) skip(report, strict, plan.typeName, gap, what);
  }
  const unknown = Object.keys(obj).filter((k) => !plan.knownFields.has(k) && !isNull(obj[k]));
  for (const key of unknown.sort()) skip(report, strict, plan.typeName, key, what);
  report.addInserted(plan.typeName);
}

function writeEntity(db: Connection, plan: ComponentPlan, obj: JsonObject) {
  const what = describe(plan.typeName, obj);
  run(db, plan.entity_sql, params(plan.entityBindings, obj, what), what);
}

function planFor(
  typeName: string,
  n: number,
  report: InsertReport,
  strict: boolean,
): ComponentPlan | undefined {
  const m = loadManifest();
  if (typeName in m.components) return m.components[typeName];
  const reason = m.unsupported_components[typeName] ?? "not a GridDB component type";
  unsupported(report, strict, typeName, n, reason);
  return undefined;
}

export function insertComponents(
  db: Connection,
  typeName: string,
  objs: JsonObject[],
  opts: InsertOptions = {},
): InsertReport {
  const strict = opts.strict ?? false;
  const report = new InsertReport();
  const plan = planFor(typeName, objs.length, report, strict);
  if (!plan) return report;
  withSavepoint(db, () => {
    for (const obj of objs) {
      writeEntity(db, plan, obj);
      writeRow(db, plan, obj, report, strict);
    }
  });
  return report;
}

export function insertComponent(
  db: Connection,
  typeName: string,
  obj: JsonObject,
  opts: InsertOptions = {},
): InsertReport {
  return insertComponents(db, typeName, [obj], opts);
}

function sectionSize(value: unknown): number {
  if (Array.isArray(value)) return value.length;
  if (value !== null && typeof value === "object") return Object.keys(value).length;
  return 0;
}

export function insertDocument(db: Connection, doc: JsonObject, opts: InsertOptions = {}): InsertReport {
  const strict = opts.strict ?? false;
  const m = loadManifest();
  const report = new InsertReport();
  const components = (doc.components ?? {}) as Record<string, JsonObject[]>;
  const plans: [ComponentPlan, JsonObject[]][] = [];
  for (const typeName of Object.keys(components).sort()) {
    const plan = planFor(typeName, components[typeName].length, report, strict);
    if (plan) plans.push([plan, components[typeName]]);
  }
  plans.sort((a, b) => a[0].rank - b[0].rank || (a[0].typeName < b[0].typeName ? -1 : 1));
  for (const section of Object.keys(m.unsupported_sections).sort()) {
    const n = sectionSize(doc[section]);
    if (n > 0) unsupported(report, strict, section, n, m.unsupported_sections[section]);
  }
  const attrTypes = new Map<number, string>();
  for (const a of (doc.supplemental_attribute_associations ?? []) as JsonObject[]) {
    attrTypes.set(a.attribute_id as number, a.attribute_type as string);
  }
  const supp = m.supplemental_attributes;
  withSavepoint(db, () => {
    seedVocabulary(db);
    for (const [plan, objs] of plans) for (const obj of objs) writeEntity(db, plan, obj);
    const routed: [string, string, JsonObject, string][] = [];
    for (const attr of (doc.supplemental_attributes ?? []) as JsonObject[]) {
      const id = attr.id as number;
      const attrType = attrTypes.get(id);
      if (attrType === undefined) {
        throw new InsertError(`supplemental attribute id=${id}: no association names its type`);
      }
      const table = supp.plant_types.includes(attrType) ? "plants" : "supplemental_attributes";
      const what = `${attrType} id=${id}`;
      run(
        db,
        "INSERT INTO entities (id, entity_table, entity_type) VALUES (?, ?, ?)",
        [BigInt(id), table, attrType],
        what,
      );
      routed.push([table, attrType, attr, what]);
    }
    for (const [plan, objs] of plans) for (const obj of objs) writeRow(db, plan, obj, report, strict);
    for (const [table, attrType, attr, what] of routed) {
      const id = BigInt(attr.id as number);
      if (table === "plants") {
        const { id: _id, name, ...rest } = attr;
        run(db, supp.plant_sql, [id, name as string, attrType, canonicalJson(rest)], what);
      } else {
        const { id: _id, ...rest } = attr;
        run(db, supp.attribute_sql, [id, attrType, canonicalJson(rest)], what);
      }
      report.addInserted(table);
    }
    for (const section of m.associations) {
      for (const row of (doc[section.section] ?? []) as JsonObject[]) {
        const what = `${section.section} row ${JSON.stringify(row)}`;
        run(db, section.row_sql, params(section.bindings, row, what), what);
        report.addInserted(section.section);
      }
    }
  });
  return report;
}
